Add scrapy-autoextract to Pypi

croqaz opened the pull request:
This PR:

* adds a few basic tests for the middleware
* use tox to test agains multiple Python versions
* adds Travis CI integration to publish tags on Pypi automatically

croqaz commented:
All tests pass, I believe it's ready to merge

kmike commented on tox.ini:
```diff
@@ -0,0 +1,10 @@
+[tox]
+envlist = py36, py37
```
Can we support Python 3.5 as well?

croqaz replied:
We should look into the future, not the past, now that Python 3.8 was released :)
But yes, I'll add Python 3.5 :)

kmike commented on README.rst:
```diff
@@ -8,6 +8,16 @@ The middleware adds the result of AutoExtract to ``response.meta['autoextract']`
+
+    pip install scrapy-autoextract
+
+scrapy-autoextract requires Python 3.
```
```suggestion
scrapy-autoextract requires Python 3.5+
```

(or 3.6+, depending on what we decide to support)

kmike commented on scrapy_autoextract/__init__.py:
```diff
@@ -0,0 +1 @@
+from .middlewares import AutoExtractMiddleware  # noqa: F401
```
👍

kmike replied:
This should allow to simplify README ("Configuration" section)

kmike commented on test/test_autoextract.py:
```diff
@@ -0,0 +1,110 @@
+from scrapy.spiders import Spider
+from scrapy.utils.test import get_crawler
+
+sys.path.append(os.getcwd())
```
why is this line needed? Tests should run for installed scrapy_autoextract. Also, os.getcwd() looks incorrect, because you may invoke these tests while the current folder is arbitrary; cwd is nott necessarily related to path to this file.

kmike commented on test/test_autoextract.py:
```diff
@@ -0,0 +1,110 @@
+def _assert_enabled(spider,
+                    settings=None,
+                    url='http://quotes.toscrape.com',
+                    proxyurl='autoextract.scrapinghub.com',
```
probably makes sense to rename "proxy..." arguments to something else, as AutoExtract is not a proxy

kmike commented on test/test_autoextract.py:
```diff
@@ -0,0 +1,110 @@
+                    settings=None,
+                    url='http://quotes.toscrape.com',
+                    proxyurl='autoextract.scrapinghub.com',
+                    proxyauth=basic_auth_header('apikey', '')):
```
this is not a value of Proxy-Auth header

kmike commented on tox.ini:
```diff
@@ -0,0 +1,10 @@
+    -r test/requirements.txt
+
+commands =
+    flake8 scrapy_autoextract
```
I think it is better to have flake8 as a separate tox environment

croqaz commented:
Ok the build for Python 3.5 is failing because of JSON and str vs bytes.
I'll install Python 3.5 tomorrow and fix it.

croqaz commented:
Fixed the tests for Python 3.5.
`json.loads(b'[{}]')` doesn't work in 3.5, but the body is required to be bytes, by the Scrapy Response object.

I believe I fixed all comments. Ready to merge 🚀

kmike commented on .travis.yml:
```diff
@@ -14,7 +14,9 @@ matrix:
 # command to run tests
-script: tox
+script:
+  - tox -e flake8
```
you can put it to matrix, as another env

croqaz replied:
Ok, I managed to do this, it took some time, but... what's the advantage? That moving this to another env will run the flake8 only once, just like a normal test?

kmike replied:
Yes; also, you won't have all tests failing if that's just a style issue. From job builds it'd be clear if this is a general issue, a problem with flake check, a problem with Python 3.5 only, etc.

kmike commented on test/test_autoextract.py:
```diff
@@ -0,0 +1,108 @@
+    assert out.headers.get('Authorization') == api_auth
+
+    resp = Response(out.url, request=out)
+    resp._body = '[{}]'  # workaround to make the body work with Python 3.5+
```
It doesn't look good; I think tests showed you a real issue in Python 3.5. In Scrapy response.body is bytes, response.text is text. In case of JSON you probably need to call `response.body.decode('utf8')`, and pass it to json.loads.

kmike commented:
thanks @croqaz!
